third_party/move/move-compiler-v2/src/cyclic_instantiation_checker.rs

```diff
@@ -0,0 +1,239 @@
+                                .inst
+                                .iter()
+                                .filter_map(ty_properly_contains_ty_parameter)
+                                .next()
+                            {
+                                self.report_error(*nid, callee, callers_chain);
+                                return false;
+                            } else {
+                                return true;
+                            }
+                        }
+                    }
+                }
+                self.visit_callees(*nid, callee, insts, callers_chain)
+            }
+        } else {
+            true
+        }
+    }
+
+    /// Visits a call expression and checks for cyclic type instantiations.
+    /// Other parameters are the same as in `visit`.
+    /// Precondition: `caller` defined in `self.mod_env`
+    fn visit_callees(
+        &self,
+        caller_node: NodeId,
+        caller: QualifiedInstId<FunId>,
+        insts: Vec<Type>,
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) -> bool {
+        let caller_body = self.get_fun_def(caller.id);
+        let caller_loc = self.mod_env.env.get_node_loc(caller_node);
+        callers_chain.push((caller_loc, caller));
+        let res = caller_body.visit_positions(&mut |pos, exp| {
+            let insts = Type::instantiate_vec(self.get_inst(caller_node), &insts);
+            self.visit(pos, exp, insts, callers_chain)
+        });
+        callers_chain.pop();
+        res.is_some()
+    }
+
+    /// Gets the funciton body
+    fn get_fun_def(&self, fun_id: FunId) -> &Exp {
+        let fun_env = self.mod_env.get_function(fun_id);
+        fun_env.data.get_def().expect("definition")
+    }
+
+    /// Shortcut for getting the node instantiation
+    fn get_inst(&self, nid: NodeId) -> Vec<Type> {
+        self.mod_env
+            .env
+            .get_node_instantiation_opt(nid)
+            .expect("instantiation")
+    }
+
+    /// Returns true if we are sure the function with given id is not recursive
+    fn def_not_recursive(&self, id: QualifiedId<FunId>) -> bool {
+        if let Some(descendants) = self
+            .mod_env
+            .env
+            .get_function(id)
+            .get_transitive_closure_of_called_functions()
+        {
+            !descendants.contains(&id)
+        } else {
+            false
+        }
+    }
+
+    /// Reports a cyclic type instantiation error, in which the root caller eventually calls `callee`
+    /// with a cyclic type instantiation.
+    fn report_error(
+        &self,
+        _nid: NodeId,
+        callee: QualifiedInstId<FunId>,
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) {
+        let labels = (0..callers_chain.len())
+            .map(|i| {
+                let (caller_loc, caller) = &callers_chain[i];
+                let callee = if i != callers_chain.len() - 1 {
```

Comment:
giving this a different name would be a good idea.  `callee_to_display` for example.  or your choice.

Reply:
Renaming to `callee_to_display` would make me want to rename `caller` to `caller_to_display` and that looks verbose. So I added some comments explaining the shadowing of `callee`.